src/Nethermind/Nethermind.Facade/Simulate/SimulateDictionaryBlockStore.cs

```diff
@@ -0,0 +1,194 @@
+// SPDX-FileCopyrightText: 2024 Demerzel Solutions Limited
+// SPDX-License-Identifier: LGPL-3.0-only
+
+using System;
+using System.Buffers;
+using System.Collections;
+using System.Collections.Generic;
+using Nethermind.Blockchain.Blocks;
+using Nethermind.Core;
+using Nethermind.Core.Buffers;
+using Nethermind.Core.Crypto;
+using Nethermind.Serialization.Rlp;
+
+namespace Nethermind.Facade.Simulate;
+
+public class SimulateDictionaryBlockStore : IBlockStore
+{
+    public class StructuralByteArrayComparer : IEqualityComparer<byte[]>
+    {
+        public bool Equals(byte[] x, byte[] y)
+        {
+            return StructuralComparisons.StructuralEqualityComparer.Equals(x, y);
+        }
+
+        public int GetHashCode(byte[] obj)
+        {
+            return StructuralComparisons.StructuralEqualityComparer.GetHashCode(obj);
+        }
+    }
+
+
+    public class CappedArrayMemoryManager : MemoryManager<byte>
+    {
+        private readonly CappedArray<byte> _data;
+        private bool _isDisposed;
+
+        public CappedArrayMemoryManager(CappedArray<byte>? data)
+        {
+            _data = data ?? throw new ArgumentNullException(nameof(data));
+            _isDisposed = false;
+        }
+
+        public override Span<byte> GetSpan()
+        {
+            if (_isDisposed)
+            {
+                throw new ObjectDisposedException(nameof(CappedArrayMemoryManager));
+            }
+
+            return _data.AsSpan();
+        }
+
+        public override MemoryHandle Pin(int elementIndex = 0)
+        {
+            if (_isDisposed)
+            {
+                throw new ObjectDisposedException(nameof(CappedArrayMemoryManager));
+            }
+
+            if (elementIndex < 0 || elementIndex >= _data.Length)
+            {
+                throw new ArgumentOutOfRangeException(nameof(elementIndex));
+            }
+            // Pinning is a no-op in this managed implementation
+            return new MemoryHandle();
+        }
+
+        public override void Unpin()
+        {
+            if (_isDisposed)
+            {
+                throw new ObjectDisposedException(nameof(CappedArrayMemoryManager));
+            }
+            // Unpinning is a no-op in this managed implementation
+        }
+
+        protected override void Dispose(bool disposing)
+        {
+            _isDisposed = true;
+        }
+
+        protected override bool TryGetArray(out ArraySegment<byte> segment)
+        {
+            if (_isDisposed)
+            {
+                throw new ObjectDisposedException(nameof(CappedArrayMemoryManager));
+            }
+
+            segment = new ArraySegment<byte>(_data.ToArray() ?? throw new InvalidOperationException());
+            return true;
+        }
+    }
+
+    private readonly IBlockStore _readonlyBaseBlockStore;
+    private readonly Dictionary<ValueHash256, Block> _blockDict = new();
+    private readonly Dictionary<long, Dictionary<Hash256, Block>> _blockNumDict = new();
```

Comment:
HashAsKey

Reply:
Do we need N:N mapping here? I think there shouldn't be multiple blocks per block number?

Reply:
used and refactored N:N